### PR TITLE
Adds support for global circuit budget

### PR DESCRIPTION
**Pull request by davinci26**
Solves #6001

### CODE REVIEW

**Comment by davinci26** [comment]
removing @clayton-gonsalves since he is on PAT leave (if you see this feel free to review anyway).

Adding @shadialtarsha and @seth-epps as well

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (davinci26) [inline apis/projectcontour/v1alpha1/contourconfig.go:116]: this requires a doc string

File: cmd/contour/servecontext.go
Comment (shadialtarsha) [inline cmd/contour/servecontext.go:591]: I would pass this as a pointer instead of relying on implicit coercion.

**Review by shadialtarsha — COMMENTED** [review]
I think the code makes sense to me. 
I have 2 questions:
1. What will happen if not all `GlobalCircuitBreakerDefaults` attributes are defined? I think they will have the value 0 and we will default to Envoy settings again which makes sense but I would clarify this in the doc.
2. Wondering if we can write e2e tests for this feature. I know it might be tricky.

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (skriss) [inline apis/projectcontour/v1alpha1/contourconfig.go:82]: I think it would make sense to move this to under `ClusterParameters` (so `spec.envoy.cluster.circuitBreakers`)

Comment (skriss) [inline apis/projectcontour/v1alpha1/contourconfig.go:82]: Let's pluralize the YAML name:
```suggestion
	// GlobalCircuitBreakerDefaults specifies default circuit breaker budget across all services.
	// If defined, this will be used as the default for all services.
	// +optional
	GlobalCircuitBreakerDefaults *GlobalCircuitBreakerDefaults `json:"circuitBreakers,omitempty"`
```

File: pkg/config/parameters.go
Comment (skriss) [inline pkg/config/parameters.go:837]: Can drop this :)

File: pkg/config/parameters.go
Comment (skriss) [inline pkg/config/parameters.go:673]: Same comments here - let's move under `ClusterParameters`, and pluralize YAML name to `circuitBreakers`.

Comment (skriss) [inline pkg/config/parameters.go:673]: Please document the new settings in https://github.com/projectcontour/contour/blob/main/site/content/docs/main/configuration.md#configuration-file as well.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1214]: I think it makes sense to have these global defaults take effect for all three processors (Ingress, HTTPProxy, Gateway) -- so need to add the implementation to `internal/dag/gatewayapi_processor.go`, and then make sure we're passing the defaults into all three processors here when we instantiate them.

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:92]: Could you explain why you want to expose this as a metric? We don't really have any other metrics like this, so it's a little unusual.

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @davinci26! Overall this makes sense to me, added some comments.

File: internal/metrics/metrics.go
Comment (davinci26) [reply internal/metrics/metrics.go:92]: > Could you explain why you want to expose this as a metric? 

We want to alert and build graphs on this as a defence in depth. We have discovered during  a deployment that we accidentally deleted those annotations and that caused issues and that was really painful to debug.

Basically it helps with the discoverability of what people are setting across the cluster.

File: cmd/contour/serve.go
Comment (davinci26) [reply cmd/contour/serve.go:1214]: Yeah I did for `ingress` and `httpproxy`, I wasnt aware if there was anything specific to gateway api spec that made this feature not that useful. Will add it

File: cmd/contour/servecontext.go
Comment (shadialtarsha) [reply cmd/contour/servecontext.go:591]: actually that is not true, both are pointers 😅

File: internal/dag/policy.go
Comment (seth-epps) [inline internal/dag/policy.go:835]: If the global circuit breaker isn't `nil`, but the fields are omitted (since they're optional), are we relying on the envoy cluster defaults?

Based on the envoy cluster configs here, https://github.com/projectcontour/contour/blob/main/internal/envoy/v3/cluster.go#L81-L90, I'm worried that any provided value will change those defaults since they're not passed as `nil`, but instead the default `uint32` 0

Comment (davinci26) [reply internal/dag/policy.go:835]: good catch!

Comment (davinci26) [reply internal/dag/policy.go:835]: Now I remember, darn I felt into this rabbit `hole` again

```
// UInt32OrNil returns a wrapped UInt32Value. If val is 0, nil is returned
func UInt32OrNil(val uint32) *wrapperspb.UInt32Value {
	switch val {
	case 0:
		return nil
	default:
		return wrapperspb.UInt32(val)
	}
}
```

handles 0 alright

File: pkg/config/parameters.go
Comment (davinci26) [reply pkg/config/parameters.go]: fixed in bdf1de3b2658964a0b571ef62674caaa8efe3ffe

Comment (davinci26) [reply pkg/config/parameters.go]: fixed in bdf1de3b2658964a0b571ef62674caaa8efe3ffe

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (davinci26) [reply apis/projectcontour/v1alpha1/contourconfig.go]: fixed in bdf1de3b2658964a0b571ef62674caaa8efe3ffe

**Comment by davinci26** [comment]
@skriss this is ready for another review round

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1214]: Still need to pass `dbc.globalCircuitBreakerDefaults` into the `IngressProcessor`, above.

File: site/content/docs/main/configuration.md
Comment (skriss) [inline site/content/docs/main/configuration.md:160]:
```suggestion
| circuit-breakers       | [CircuitBreakers](#circuit-breakers)    | none    | This field specifies the default value for [circuit-breaker-annotations](https://projectcontour.io/docs/main/config/annotations/) for services that don't specify them.                                                                    |
```

File: changelogs/unreleased/6013-davinci26-small.md
Comment (skriss) [inline changelogs/unreleased/6013-davinci26-small.md:1]: Let's bump this up to `minor` since it has a header and more than a sentence of description, so need to rename the file and update the PR label.

File: changelogs/unreleased/6013-davinci26-small.md
Comment (skriss) [inline changelogs/unreleased/6013-davinci26-small.md:4]: Let's mention that these are set in the config file via `cluster.circuit-breakers`, and add a link to https://projectcontour.io/docs/1.28/configuration/#circuit-breakers release, or via ContourConfiguration CRD in `spec.envoy.cluster.circuitBreakers` with a link to https://projectcontour.io/docs/1.28/config/api/#projectcontour.io/v1alpha1.GlobalCircuitBreakerDefaults

It's fine that the links are broken right now, they'll resolve when the release is cut.

**Review by skriss — CHANGES_REQUESTED** [review]

File: cmd/contour/serve.go
Comment (davinci26) [reply cmd/contour/serve.go:1214]: thanks nice catch, I fixed the `cluster_test` and I had a false sense of confidence, I codified this in `serve_test` as well

File: changelogs/unreleased/6013-davinci26-small.md
Comment (davinci26) [reply changelogs/unreleased/6013-davinci26-small.md]: fixed in 4151035aa07d44ff5b162b52dd73935da4f7cda2

File: changelogs/unreleased/6013-davinci26-small.md
Comment (davinci26) [reply changelogs/unreleased/6013-davinci26-small.md]: fixed in 4151035aa07d44ff5b162b52dd73935da4f7cda2

File: site/content/docs/main/configuration.md
Comment (davinci26) [reply site/content/docs/main/configuration.md]: fixed in 4151035aa07d44ff5b162b52dd73935da4f7cda2

File: cmd/contour/serve.go
Comment (davinci26) [reply cmd/contour/serve.go:1215]: fixed in 4151035aa07d44ff5b162b52dd73935da4f7cda2 + added tests

**Comment by davinci26** [comment]
@skriss unless there is an embarrassing bug somewhere this is ready

**Comment by skriss** [comment]
Looks like we got some merge conflicts, if you can resolve those I'll take another look

**Comment by davinci26** [comment]
@skriss merge conflicts resolved

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:247]: Help text should be updated here to be relevant

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:245]: Please also add the new metric to the `Zero()` function and re-run `make generate` to generate docs for it.

**Review by skriss — APPROVED** [review]
This LGTM, I'm approving with comments that need to be addressed but can be done in a follow-up PR if preferred. Will leave for another maintainer to take a look as well. Thanks @davinci26!

File: internal/metrics/metrics.go
Comment (davinci26) [reply internal/metrics/metrics.go]: yeah yeah you right, fixed now!

File: internal/metrics/metrics.go
Comment (davinci26) [reply internal/metrics/metrics.go:245]: Cool I didn't know about this

**Comment by davinci26** [comment]
@projectcontour/maintainers can someone PTAL as well

File: internal/metrics/metrics.go
Comment (sunjayBhatia) [reply internal/metrics/metrics.go:92]: Hm, I would expect something like this to be exposed via resource Status or something similar rather than a Contour xDS server metric

Also I believe we have an issue/PR to expand on cluster circuit breaker metrics: https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#circuit-breakers-statistics (here: https://github.com/projectcontour/contour/pull/5950)

I don't want to hold up the rest of the change, so maybe lets pull the metrics part out into another PR?

Comment (davinci26) [reply internal/metrics/metrics.go:92]: > Hm, I would expect something like this to be exposed via resource Status or something similar rather than a Contour xDS server metric


Personally I take a more liberal approach to metrics in the sense that, if `contour` makes the decision to set a specific value it would be useful for operators to know the decision at runtime, especially if it affect things

Comment (davinci26) [reply internal/metrics/metrics.go:92]: > I don't want to hold up the rest of the change, so maybe lets pull the metrics part out into another PR?

I won't die on this hill :D, So I  can separate the changes into two PRs. But my general take is that adding the metric is pretty "cheap" and overall useful. I.e. there have been cases where I wish I had this metric tracked.

**Review by sunjayBhatia — COMMENTED** [review]
i think everything LGTM minus the metric part, I wonder if we can pull that out into a separate PR so we can discuss a little more?

**Comment by davinci26** [comment]
> i think everything LGTM minus the metric part, I wonder if we can pull that out into a separate PR so we can discuss a little more?

Will do it in a separate PR

**Comment by davinci26** [comment]
there is more cleanup to be made

**Comment by davinci26** [comment]
I think I got all the references now

File: internal/metrics/metrics.go
Comment (sunjayBhatia) [inline internal/metrics/metrics.go:91]: last thing to remove i think here

File: site/content/docs/v1.9.0/annotations.md
Comment (sunjayBhatia) [inline site/content/docs/v1.9.0/annotations.md:26]: can probably get rid of this change being part of this PR too, since this is quite an old version etc.

**Review by sunjayBhatia — APPROVED** [review]

**Comment by therealak12** [comment]
Also, solves https://github.com/projectcontour/contour/issues/1192. Doesn't it?

**Comment by skriss** [comment]
> Also, solves #1192. Doesn't it?

Looks like it, closing that one out.

**Pull request merged**
